subsys/bluetooth/mesh/crypto_mtls.c

```diff
@@ -213,12 +227,124 @@ int bt_mesh_aes_cmac(const uint8_t key[16], struct bt_mesh_sg *sg,
 	return err;
 }
 
+int bt_mesh_pub_key_gen(void)
+{
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	psa_status_t status;
+	int err = 0;
+	size_t key_len;
+
+	psa_destroy_key(key.priv_key_id);
+	key.is_ready = false;
+
+	/* Crypto settings for ECDH using the SHA256 hashing algorithm,
+	 * the secp256r1 curve
+	 */
+	psa_set_key_usage_flags(&key_attributes, PSA_KEY_USAGE_DERIVE);
+	psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+	psa_set_key_algorithm(&key_attributes, PSA_ALG_ECDH);
+	psa_set_key_type(&key_attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_bits(&key_attributes, 256);
+
+	/* Generate a key pair */
+	status = psa_generate_key(&key_attributes, &key.priv_key_id);
+	if (status != PSA_SUCCESS) {
+		err = -EIO;
+		goto end;
+	}
+
+	status = psa_export_public_key(key.priv_key_id, key.public_key_be,
+				sizeof(key.public_key_be), &key_len);
+	if (status != PSA_SUCCESS) {
+		err = -EIO;
+		goto end;
+	}
+
+	if (key_len != PUB_KEY_SIZE + 1) {
+		err = -ERANGE;
+		goto end;
+	}
+
+	key.is_ready = true;
+
+end:
+	psa_reset_key_attributes(&key_attributes);
+
+	return err;
+}
+
+const uint8_t *bt_mesh_pub_key_get(void)
+{
+	return key.is_ready ? key.public_key_be + 1 : NULL;
+}
+
+BUILD_ASSERT(PSA_RAW_KEY_AGREEMENT_OUTPUT_SIZE(
+	PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1), 256) == DH_KEY_SIZE,
+	"Diffie-Hellman shared secret size should be the same in PSA and BLE Mesh");
+
+BUILD_ASSERT(PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE(256) == PUB_KEY_SIZE + 1,
+	"Exported PSA public key should be 1 byte larger than BLE Mesh public key");
+
 int bt_mesh_dhkey_gen(const uint8_t *pub_key, const uint8_t *priv_key, uint8_t *dhkey)
 {
-	return -ENOSYS;
+	int err = 0;
+	psa_key_id_t priv_key_id  = PSA_KEY_ID_NULL;
+	uint8_t public_key_repr[PUB_KEY_SIZE + 1];
+	psa_status_t status;
+	size_t dh_key_len;
+
+	if (priv_key) {
+		psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+
+		/* Import a custom private key */
+		psa_set_key_usage_flags(&attributes, PSA_KEY_USAGE_DERIVE);
+		psa_set_key_lifetime(&attributes, PSA_KEY_LIFETIME_VOLATILE);
+		psa_set_key_algorithm(&attributes, PSA_ALG_ECDH);
+		psa_set_key_type(&attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+		psa_set_key_bits(&attributes, 256);
+
+		status = psa_import_key(&attributes, priv_key, PRIV_KEY_SIZE, &priv_key_id);
+		if (status != PSA_SUCCESS) {
+			err = -EIO;
+			goto end;
+		}
+
+		psa_reset_key_attributes(&attributes);
+	} else {
+		priv_key_id = key.priv_key_id;
```

Comment:
You cannot call at first without `bt_mesh_pub_key_gen` and `bt_mesh_pub_key_get` since you do not have public key :)
Getting public key returns NULL.